plugins/circleci/src/hooks/usePipelines.ts

```diff
@@ -0,0 +1,137 @@
+import { useProjectSlugFromEntity } from './useProjectSlugFromEntity';
+import { circleCIApiRef } from '../api';
+
+type PipelinesPage = {
+  items?: PipelineInfo[];
+  pageToken?: string;
+  nextPageToken?: string;
+};
+
+export function usePipelines() {
+  const { repo, owner, projectSlug } = useProjectSlugFromEntity();
+  const api = useApi(circleCIApiRef);
+  const errorApi = useApi(errorApiRef);
+  const pipelinesPage = useRef<PipelinesPage>({ items: [] });
+  const [pageToken, setPageToken] = useState<string>();
+  const [pipelines, setPipelines] = useState<PipelineInfo[]>([]);
+
+  const getWorkflows = useCallback(
+    async (pipelineId: string) => {
+      try {
+        const response = await api.getWorkflowsForPipeline(pipelineId);
+        return response.items;
+      } catch (e) {
+        if (!(e instanceof AuthenticationError)) {
+          errorApi.post(e);
+        }
+        return Promise.reject(e);
+      }
+    },
+    [api, errorApi],
+  );
+
+  const transform = async (
+    pipelinesData: Pipeline[],
+    getWorkflowsCall: { (pipelineId: string): Promise<Workflow[]> },
+  ): Promise<PipelineInfo[]> => {
+    const pipelinePromises = pipelinesData.map(async pipelineData => {
+      const workflows = await getWorkflowsCall(pipelineData.id);
+      const tableBuildInfo: PipelineInfo = {
+        ...pipelineData,
+        workflows,
+      };
+      return tableBuildInfo;
+    });
+
+    return Promise.all(pipelinePromises);
+  };
+
+  const getPipelines = useCallback(async (): Promise<PipelineInfo[]> => {
+    if (!projectSlug) {
+      return Promise.reject(new Error('Missing project slug'));
+    }
+
+    const { items = [] } = pipelinesPage.current;
+
+    try {
+      const response = await api.getPipelinesForProject(projectSlug, pageToken);
+
+      const allPipelines = [
+        ...items,
+        ...(await transform(response.items ?? [], getWorkflows)),
+      ];
+
+      pipelinesPage.current = {
+        items: allPipelines,
+        nextPageToken: response.next_page_token,
+      };
+
+      return allPipelines;
+    } catch (e) {
+      if (!(e instanceof AuthenticationError)) {
+        errorApi.post(e);
+      }
+      return Promise.reject(e);
+    }
+  }, [projectSlug, pageToken, getWorkflows, api, errorApi]);
+
+  const { value, loading, retry } = useAsyncRetry(() => {
+    return getPipelines();
+  }, [getPipelines]);
+
+  useEffect(() => setPipelines(value || []), [value]);
+
+  const rerunWorkflow = async (workflowId: string) => {
+    try {
+      await api.rerunWorkflow(workflowId);
+    } catch (e) {
+      errorApi.post(e);
+    }
+  };
+
+  const fetchMore = () => {
+    setPageToken(pipelinesPage.current.nextPageToken);
+  };
+
+  const reload = () => {
+    pipelinesPage.current = {};
+    setPipelines([]);
+    retry();
+  };
+
+  const projectName = `${owner}/${repo}`;
+
+  return [
+    {
+      loading,
+      pipelines: pipelines || [],
```

Comment:
note that this creates a new array every time if you are in the fallback case; you may want to return that from inside the useAsyncRetry above instead

Reply:
Fixed